Sources/SwiftInspectorCommands/TypeLocationCommand.swift

```diff
@@ -0,0 +1,80 @@
+// Created by Michael Bachand on 3/28/20.
+//
+// Copyright (c) 2020 Francisco Diaz
+//
+// Distributed under the MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import ArgumentParser
+import Foundation
+import SwiftInspectorCore
+
+final class TypeLocationCommand: ParsableCommand {
+  static var configuration = CommandConfiguration(
+    commandName: "type-location",
+    abstract: "Finds the location of a type"
+  )
+
+  @Option(help: nameArgumentHelp)
+  var name: String
+
+  @Option(help: "The absolute path to the file to inspect")
+  var path: String
+
+  /// Runs the command
+  func run() throws {
+    let cachedSyntaxTree = CachedSyntaxTree()
+
+    let analyzer = TypeLocationAnalyzer(typeName: name, cachedSyntaxTree: cachedSyntaxTree)
+    let fileURL = URL(fileURLWithPath: path)
+    let typeLocations = try analyzer.analyze(fileURL: fileURL)
+
+    if !typeLocations.isEmpty {
+      print(typeLocations.map { $0.outputString() }.joined(separator: "\n"))
+    }
+  }
+
+  /// Validates if the arguments of this command are valid
+  func validate() throws {
+    guard !name.isEmpty else {
+      throw InspectorError.emptyArgument(argumentName: "--name")
+    }
+    guard !path.isEmpty else {
+      throw InspectorError.emptyArgument(argumentName: "--path")
+    }
+    var isDir: ObjCBool = false
+    guard FileManager.default.fileExists(atPath: path, isDirectory: &isDir) && !isDir.boolValue else {
+      throw InspectorError.invalidArgument(argumentName: "--path", value: path)
+    }
+  }
+}
+
+extension LocatedType {
```

Comment:
The reason why we're not extending the type that's outputted from the Analyzer in other places is that this won't work as cleanly when we need to have different output types. I actually started with this path and moved away from it when that use case came up for the imports command.

To allow that use case with this pattern, we'd need to pass the output type parameter from the Command to this type. I've been trying to keep a clear separation of types between the Command and the Analyzer types, where the Analyzer types shouldn't know anything "upwards", but it's ok to have the Command know about the Analyzer (downwards) since the Command uses the Analyzer. This would break that contract and would make it somewhat harder to extend when we have more output types.


So the pattern I've been using is to have the command (which already is responsible for input/output) to be the one that manipulates the LocatedType and convert that into output. That way the LocatedType doesn't know anything about commands, but the command keeps being responsible from input/output.

You can read more about [this here](https://github.com/fdiaz/SwiftInspector/blob/master/CONTRIBUTING.md#structure-of-swiftinspector)

Reply:
https://github.com/fdiaz/SwiftInspector/pull/13/commits/4b2f3e3b36d198cf222336fc663836320f289cd6